### PR TITLE
Fix iterator position calculation for array patterns

### LINKED ISSUE #3455: Assertion ecma_is_value_object (iterator) in ecma_op_iterator_next
###### JerryScript revision
d0b4e7dd

###### Build platform
Linux-4.15.0-72-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
[...RegExp.$.$] = String()
```

###### Output
```text
ICE: Assertion 'ecma_is_value_object (iterator)' failed at jerryscript/jerry-core/ecma/operations/ecma-iterator-object.c(ecma_op_iterator_next):246.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x5657b7cd in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x5664976b in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x566497ac in jerry_assert_fail (assertion=0x566ed7e0 "ecma_is_value_object (iterator)", file=0x566ed5c0 "jerryscript/jerry-core/ecma/operations/ecma-iterator-object.c", function=0x566b3760 <__func__.6274.lto_priv.639> "ecma_op_iterator_next", line=246) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x56667ca2 in ecma_op_iterator_next (iterator=2221, value=8) at jerryscript/jerry-core/ecma/operations/ecma-iterator-object.c:246
#7  0x566682d1 in ecma_op_iterator_step (iterator=2221) at jerryscript/jerry-core/ecma/operations/ecma-iterator-object.c:509
#8  0x565d6bda in vm_loop (frame_ctx_p=0xffffc6c0) at jerryscript/jerry-core/vm/vm.c:2007
#9  0x565e29e3 in vm_execute (frame_ctx_p=0xffffc6c0) at jerryscript/jerry-core/vm/vm.c:4070
#10 0x565e3063 in vm_run (bytecode_header_p=0xf4f01ea0, this_binding_value=4126149459, lex_env_p=0xf5d007b0, arg_list_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:4199
#11 0x5662d02e in vm_run_global (bytecode_p=0xf4f01ea0) at jerryscript/jerry-core/vm/vm.c:286
#12 0x5669ab1f in jerry_run (func_val=4126148883) at jerryscript/jerry-core/api/jerry.c:595
#13 0x5669728c in main (argc=2, argv=0xffffcb14) at jerryscript/jerry-main/main-unix.c:740
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### PR DESCRIPTION
**Pull request by rerobika**
This patch fixes #3455.

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-parser-expr.c [hunk 1]
```diff
@@ -2608,8 +2608,11 @@ parser_pattern_process_assignment (parser_context_t *context_p, /**< context */
   {
     parser_parse_expression (context_p, PARSE_EXPR_NO_COMMA | PARSE_EXPR_LEFT_HAND_SIDE);
 
+
```
Comment (galpeter) [inline jerry-core/parser/js/js-parser-expr.c]: I don't think we need this extra line.

File: jerry-core/parser/js/js-parser-expr.c [hunk 2]
```diff
@@ -2608,8 +2608,11 @@ parser_pattern_process_assignment (parser_context_t *context_p, /**< context */
...
     if (!PARSER_IS_PUSH_LITERAL (context_p->last_cbc_opcode))
     {
+      uint16_t opcode_inc = 2;
```
Comment (galpeter) [inline jerry-core/parser/js/js-parser-expr.c]: This could be moved below after the raise error and just before the `last_cbc_opcode` check.

Comment (galpeter) [reply jerry-core/parser/js/js-parser-expr.c]: And.... Why `2`? Would it be possible to have a bit of comment here on this?

File: jerry-core/vm/vm.c [hunk 3]
```diff
@@ -1948,8 +1948,8 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
         }
         case VM_OC_ITERATOR_STEP:
         {
-          const int8_t index = (opcode == CBC_EXT_ITERATOR_STEP) ? -1 : -3;
-          result = ecma_op_iterator_step (stack_top_p[index]);
+          const uint8_t index = (uint8_t) (1 + (opcode - CBC_EXT_ITERATOR_STEP));
```
Comment (galpeter) [inline jerry-core/vm/vm.c]: Is there a valid range of indexes we support? If so would it be possible to have an assert for it?

File: jerry-core/vm/vm.c [hunk 4]
```diff
@@ -1999,9 +1999,9 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
         }
         case VM_OC_REST_INITIALIZER:
         {
-          const int8_t iterator_index = (opcode == CBC_EXT_REST_INITIALIZER) ? -1 : -3;
+          const uint8_t iterator_index = (uint8_t) (1 + (opcode - CBC_EXT_REST_INITIALIZER));
```
Comment (galpeter) [inline jerry-core/vm/vm.c]: Same question as above.

File: jerry-core/vm/vm.c (hunk 3 above)
Comment (rerobika) [reply jerry-core/vm/vm.c]: Yes, there is. Good idea, I'll add an assert for it.

**Comment by rerobika** [comment]
@galpeter Thanks for the review I've updated the PR. Please take an other look at it.

**Pull request merged**